### ISSUE TITLE
Units are stuck between buildings

### ISSUE THREAD

**Issue by RaphaelRoyerRivard** [RaphaelRoyerRivard/MicroMachine#877]
https://ai-arena.net/matches/210474/
![image](https://user-images.githubusercontent.com/5833235/80283630-0f457900-86e7-11ea-99f9-0c0b73a1f9b2.png)

**Comment by Unusual229** [comment]
Holy! This is SO specific.
It cannot be fixed by reserving more tiles, would just make the building pattern  different.
Two possible solutions:
1- Before placing any building, we have to pathfind from all the producers to see if the unit will be stuck or not. Will likely be resource intensive.
2- Detect stuck unit, when it happens we need to move a building.

**Comment by Unusual229** [comment]
We could also do a much simpler detection of stuck units which is to path find whenever a new unit is created. So for each unit created in the main (so not in the proxy), path find to the natural, if it fails the unit is stuck.

To choose the building to lift, we want to check if any of them doesn't have an addon yet, pick that one. Otherwise, the first one to the idle should lift. The tiles where the building used to be need to remain reserved.

**Comment by Unusual229** [comment]
There is an even simpler solution. I actually tried (in paint) to add more reserved tiles and see if it would be possible to created a situation where units are blocked.
I found that by adding 2 tiles to the left of the production building, it makes is extremely hard for units to be stuck. It requires a combination of non-production building (armory, engeneering bay, cc), Starports (do not have any reserved tiles around it) and/or unwalkable terrain.
